### ISSUE TITLE
[glslang] import failure

### ISSUE THREAD

**Issue by qbojj** [microsoft/vcpkg#27121]
**Host Environment**

- OS: Linux (Github CI ubuntu-latest)
- Compiler: GNU 9.4.0

**To Reproduce**

After successful install:
call find_package( glslang REQUIRED )
without prior call to find_package( Threads )

**Failure logs**

```
   Preset CMake variables:
  
    CMAKE_INSTALL_PREFIX:PATH="/home/runner/work/4dGraphics/4dGraphics/out/install/4DG.app"
    CMAKE_TOOLCHAIN_FILE:FILEPATH="/home/runner/work/4dGraphics/4dGraphics/vcpkg/scripts/buildsystems/vcpkg.cmake"
  
  -- Running vcpkg install
  Detecting compiler hash for triplet x64-linux...
  The following packages will be built and installed:
      assimp[core]:x64-linux -> 5.2.5
    * draco[core]:x64-linux -> 1.5.2
      glfw3[core]:x64-linux -> 3.3.8
      glm[core]:x64-linux -> 0.9.9.8#1
      glslang[core]:x64-linux -> 11.11.0
      imgui[core,glfw-binding]:x64-linux -> 1.88#1
    * kubazip[core]:x64-linux -> 0.2.4
    * minizip[core]:x64-linux -> 1.2.11#11
    * poly2tri[core]:x64-linux -> 2020-07-21#2
      portable-snippets[core]:x64-linux -> 2019-09-20#2
    * pugixml[core]:x64-linux -> 1.12.1#1
    * rapidjson[core]:x64-linux -> 2022-06-28#3
      stb[core]:x64-linux -> 2021-09-10#1
    * utfcpp[core]:x64-linux -> 3.2.1#1
    * vcpkg-cmake[core]:x64-linux -> 2022-08-18
    * vcpkg-cmake-config[core]:x64-linux -> 2022-02-06#1
      volk[core]:x64-linux -> 1.3.204#1
    * vulkan[core]:x64-linux -> 1.1.82.1#2
      vulkan-memory-allocator[core]:x64-linux -> 3.0.0
    * zlib[core]:x64-linux -> 1.2.12#2
  Additional packages (*) will be modified to complete this operation.
  Restored 20 package(s) from /home/runner/work/4dGraphics/b/vcpkg_cache in 3.525 s. Use --debug to see more details.
  Installing 1/20 vcpkg-cmake:x64-linux...
  Elapsed time to handle vcpkg-cmake:x64-linux: 3.794 ms
  Installing 2/20 zlib:x64-linux...
  Elapsed time to handle zlib:x64-linux: 22.26 ms
  Installing 3/20 vcpkg-cmake-config:x64-linux...
  Elapsed time to handle vcpkg-cmake-config:x64-linux: 2.455 ms
  Installing 4/20 utfcpp:x64-linux...
  Elapsed time to handle utfcpp:x64-linux: 26.86 ms
  Installing 5/20 stb:x64-linux...
  Elapsed time to handle stb:x64-linux: 5.425 ms
  Installing 6/20 rapidjson:x64-linux...
  Elapsed time to handle rapidjson:x64-linux: 5.591 ms
  Installing 7/20 pugixml:x64-linux...
  Elapsed time to handle pugixml:x64-linux: 3.912 ms
  Installing 8/20 poly2tri:x64-linux...
  Elapsed time to handle poly2tri:x64-linux: 4.31 ms
  Installing 9/20 minizip:x64-linux...
  Elapsed time to handle minizip:x64-linux: 3.29 ms
  Installing 10/20 kubazip:x64-linux...
  Elapsed time to handle kubazip:x64-linux: 6.844 ms
  Installing 11/20 draco:x64-linux...
  Elapsed time to handle draco:x64-linux: 524 ms
  Installing 12/20 assimp:x64-linux...
  Elapsed time to handle assimp:x64-linux: 2.092 s
  Installing 13/20 glfw3:x64-linux...
  Elapsed time to handle glfw3:x64-linux: 6.303 ms
  Installing 14/20 glm:x64-linux...
  Elapsed time to handle glm:x64-linux: 618.3 ms
  Installing 15/20 glslang:x64-linux...
  Elapsed time to handle glslang:x64-linux: 138.6 ms
  Installing 16/20 imgui:x64-linux...
  Elapsed time to handle imgui:x64-linux: 10.55 ms
  Installing 17/20 portable-snippets:x64-linux...
  Elapsed time to handle portable-snippets:x64-linux: 220.1 ms
  Installing 18/20 vulkan:x64-linux...
  Elapsed time to handle vulkan:x64-linux: 1.666 ms
  Installing 19/20 volk:x64-linux...
  Elapsed time to handle volk:x64-linux: 4.365 ms
  Installing 20/20 vulkan-memory-allocator:x64-linux...
  Elapsed time to handle vulkan-memory-allocator:x64-linux: 3.964 ms
  
  Total elapsed time: 7.902 s
  stb is header-only and can be used from CMake via:
  
      find_path(STB_INCLUDE_DIRS "stb_c_lexer.h")
      target_include_directories(main PRIVATE ${STB_INCLUDE_DIRS})
  
  assimp provides CMake targets:
  
      # this is heuristically generated, and may not be correct
      find_package(assimp CONFIG REQUIRED)
      target_link_libraries(main PRIVATE assimp::assimp)
  
  glfw3 provides CMake targets:
  
      # this is heuristically generated, and may not be correct
      find_package(glfw3 CONFIG REQUIRED)
      target_link_libraries(main PRIVATE glfw)
  
  glm provides CMake targets:
  
      # this is heuristically generated, and may not be correct
      find_package(glm CONFIG REQUIRED)
      target_link_libraries(main PRIVATE glm::glm)
  
  The package glslang provides CMake targets:
  
      find_package(glslang CONFIG REQUIRED)
      target_link_libraries(main PRIVATE glslang::OSDependent glslang::glslang glslang::MachineIndependent glslang::GenericCodeGen glslang::OGLCompiler glslang::glslangValidator glslang::spirv-remap glslang::glslang-default-resource-limits glslang::SPVRemapper glslang::SPIRV glslang::HLSL)
  
  imgui provides CMake targets:
  
      # this is heuristically generated, and may not be correct
      find_package(imgui CONFIG REQUIRED)
      target_link_libraries(main PRIVATE imgui::imgui)
  
  portable-snippets provides CMake targets:
  
      # this is heuristically generated, and may not be correct
      find_package(unofficial-portable-snippets CONFIG REQUIRED)
      # note: 8 additional targets are not displayed.
      target_link_libraries(main PRIVATE unofficial::portable-snippets::cpu unofficial::portable-snippets::once unofficial::portable-snippets::check unofficial::portable-snippets::clock)
  
  volk provides CMake targets:
  
      # this is heuristically generated, and may not be correct
      find_package(volk CONFIG REQUIRED)
      target_link_libraries(main PRIVATE volk::volk volk::volk_headers)
  
  vulkan-memory-allocator provides CMake targets:
  
      # this is heuristically generated, and may not be correct
      find_package(unofficial-vulkan-memory-allocator CONFIG REQUIRED)
      target_link_libraries(main PRIVATE unofficial::vulkan-memory-allocator::vulkan-memory-allocator)
  
  -- Running vcpkg install - done
  -- The C compiler identification is GNU 9.4.0
  -- The CXX compiler identification is GNU 9.4.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Optick is disabled
  -- Found ZLIB: optimized;/home/runner/work/4dGraphics/4dGraphics/out/build/vcpkg_installed/x64-linux/lib/libz.a;debug;/home/runner/work/4dGraphics/4dGraphics/out/build/vcpkg_installed/x64-linux/debug/lib/libz.a (found version "1.2.12") 
  -- RapidJSON found. Headers: /home/runner/work/4dGraphics/4dGraphics/out/build/vcpkg_installed/x64-linux/share/rapidjson/../../include
  -- Found Stb: /home/runner/work/4dGraphics/4dGraphics/out/build/vcpkg_installed/x64-linux/include  
  -- Configuring done
  CMake Error at out/build/vcpkg_installed/x64-linux/share/glslang/glslang-targets.cmake:60 (set_target_properties):
  Error:   The link interface of target "glslang::OSDependent" contains:
  
      Threads::Threads
  
    but the target was not found.  Possible reasons include:
  
      * There is a typo in the target name.
      * A find_package call is missing for an IMPORTED target.
      * An ALIAS target is missing.
  
  Call Stack (most recent call first):
    out/build/vcpkg_installed/x64-linux/share/glslang/glslang-config.cmake:26 (include)
    vcpkg/scripts/buildsystems/vcpkg.cmake:833 (_find_package)
    shared/CMakeLists.txt:16 (find_package)
  
  
  -- Generating done
  CMake Generate step failed.  Build files cannot be regenerated correctly.
⏱ elapsed: 8.885 seconds
Error: "CMake failed with error code: '1'.
    at CMakeRunner.<anonymous> (/home/runner/work/_actions/lukka/run-cmake/v10/dist/index.js:4164:23)
    at Generator.next (<anonymous>)
    at fulfilled (/home/runner/work/_actions/lukka/run-cmake/v10/dist/index.js:4007:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
Error: run-cmake action execution failed: 'Error: "CMake failed with error code: '1'.'
```

**Additional context**
To circumvent just call find_package( Threads ) beforehand

It's not really a build failure but a import failure and I don't know where to submit such cases

**Comment by JonLiu1993** [comment]
@qbojj, Thanks for posting this issue, after add find_package( Threads ), can you successfully use glslang?

**Comment by qbojj** [comment]
Yes I, after adding it before find_package( glslang ) I can use it.

**Comment by iaomw** [comment]
It's telling me `glslang::glslangValidator` and `glslang::spirv-remap` are missing.
```
-- Configuring done
CMake Error at /home/iaomw/Github/vcpkg/scripts/buildsystems/vcpkg.cmake:582 (_add_executable):
  Target "vktest" links to target "glslang::glslangValidator" but the target
  was not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:20 (add_executable)


CMake Error at /home/iaomw/Github/vcpkg/scripts/buildsystems/vcpkg.cmake:582 (_add_executable):
  Target "vktest" links to target "glslang::spirv-remap" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:20 (add_executable)


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```
CMake contents:

```
set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
set(THREADS_PREFER_PTHREAD_FLAG TRUE)
find_package(Threads REQUIRED)

find_package(glslang CONFIG REQUIRED)
target_link_libraries(${PROJECT_NAME}
                                      glslang::OSDependent 
                                      glslang::glslang 
                                      glslang::MachineIndependent 
                                      glslang::GenericCodeGen 
                                      glslang::OGLCompiler 
                                      glslang::glslangValidator 
                                      glslang::spirv-remap 
                                      glslang::glslang-default-resource-limits 
                                      glslang::SPVRemapper 
                                      glslang::SPIRV 
                                      glslang::HLSL
                                      )
```

**Comment by JonLiu1993** [comment]
@qbojj, I found the same [issue](https://discourse.cmake.org/t/cmake-3-21-rc-1-threads-threads-not-found/3630) as you in the cmake official community. It seems that this is related to cmake. The solution they gave is to add find_package( Threads ) when using.

**Comment by JonLiu1993** [comment]
@iaomw, Are you getting the error on windows or linux environment?

**Comment by JonLiu1993** [comment]
@qbojj @iaomw, the PR https://github.com/microsoft/vcpkg/pull/30374 update glslang to new version 12.1.0, I tested usage by `glslang:x64-windows` successfully on my windows machine and use VS2019.
````
>------ Build All started: Project: CMakeProject42, Configuration: x64-Debug ------
  Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
  Copyright (C) Microsoft Corporation. All rights reserved.
  
    Checking Build System
    Building Custom Rule C:/Users/test/source/repos/CMakeProject42/CMakeProject42/CMakeLists.txt
    CMakeProject42.cpp
    main.vcxproj -> C:\Users\test\source\repos\CMakeProject42\out\build\x64-Debug\CMakeProject42\Debug\main.exe
    Building Custom Rule C:/Users/test/source/repos/CMakeProject42/CMakeLists.txt

Build All succeeded.
````
After this PR merge, you can use the latest vcpkg to get the latest version of glslang and try it out.

**Comment by moritz-h** [comment]
@JonLiu1993 (Coming here from your mention of this in #30374.)
This issue here will not be fixed by 12.1.0. This behaviour is consistent with upstream and only affects Linux/Unix as it comes from `glslang::OSDependent`.
Here is the corresponding upstream issue: https://github.com/KhronosGroup/glslang/issues/3125

**Comment by moritz-h** [comment]
Made an upstream PR, so hopefully will be fixed in 12.2.0 someday.
